h5py/_hl/files.py

```diff
@@ -342,15 +343,20 @@ def __init__(self, name, mode=None, driver=None,
             with phil:
                 fid = h5i.get_file_id(name)
         else:
-            name = filename_encode(name)
+            if hasattr(name, 'read') and hasattr(name, 'seek'):
```

Comment:
`write`, `truncate`, and `flush` -- only occasionally. It should really come quite natural for any h5py user that fileobj *must* implement those methods in order to write data.
I added `read` and `seek` safe-guards to make code more readable and self-documented, not to prevent all misuses (in Python, that's impossible). Technically more checks could be added of course -- but is it worth the complexity?